Active Storage: docs on lazy vs immediate loading (#42495)
rails/rails: guides/source/active_storage_overview.md

[`representable?`]: https://api.rubyonrails.org/classes/ActiveStorage/Blob/Representable.html#method-i-representable-3F
[`representation`]: https://api.rubyonrails.org/classes/ActiveStorage/Blob/Representable.html#method-i-representation

> ### Lazy vs Immediate Loading

Reviewer: I think loading is a bit ambiguous here.
Maybe:
```suggestion
### Lazy vs Immediate Processing
```

Author: I think "lazy loading" is a much more commonly used term to describe the pattern.

![image](https://user-images.githubusercontent.com/509837/122248215-8ca0ef80-ce8d-11eb-973a-15ac4396b339.png)
![image](https://user-images.githubusercontent.com/509837/122248264-9591c100-ce8d-11eb-9370-5d1cba244e7d.png)

Reviewer: Ok, but in this case it's mostly about deferring the processing to another request right? At least the first time, when blobs haven't been processed...

> By default, Active Storage will represent files lazily. This code:

Reviewer: ```suggestion
By default, Active Storage will process representations lazily. This code:
```

```ruby
image_tag file.representation(resize_to_limit: [100, 100])
```

Will generate an `<img>` tag with the `src` pointing to the [`ActiveStorage::Representations::RedirectController`][].
The browser will make a request to that controller, which will return a `302` redirect to the file on the remote service (or in [proxy mode](#proxy-mode), return the file contents). Loading the file lazily allows features like [single use URLs](#public-access) to work without slowing down your initial page loads.

This works fine for most cases.

If you want to generate URLs for images immediately, you can call `.processed.url`:

```ruby
image_tag file.representation(resize_to_limit: [100, 100]).processed.url
```

> The Active Storage variant tracker improves performance of this, by storing a record in the database if the requested representation has happened before. Thus, the above code will only make an API call to the remote service (eg. S3) once, and once a variant is stored, will use that. The variant tracker runs automatically, but can be disabled through `config.active_storage.track_variants`.

Reviewer: ```suggestion
The Active Storage variant tracker improves performance of this, by storing a record in the database if the requested representation has been process before. Thus, the above code will only make an API call to the remote service (eg. S3) once, and once a variant is stored, will use that. The variant tracker runs automatically, but can be disabled through `config.active_storage.track_variants`.
```

Reviewer: Could you also wrap all lines around 80 characters? (This should probably be added to the Contributing Guide)

Author: I'm not sure how it makes the docs more readable, but happy to do it since it is the standard I guess.

> The Active Storage variant tracker improves performance of this, by storing a record in the database of if the requested representation has happened before. Thus, the above code will only make an API call to the remote service (eg. S3) once, and once a variant is stored, will use that. The variant tracker runs automatically, but can be disabled through `config.active_storage.track_variants`.

Reviewer: ```suggestion
The Active Storage variant tracker improves performance of this, by storing a record in the database if the requested representation has happened before. Thus, the above code will only make an API call to the remote service (eg. S3) once, and once a variant is stored, will use that. The variant tracker runs automatically, but can be disabled through `config.active_storage.track_variants`.
```

> If you're rendering lots of images on a page, the above example could result in N+1 queries loading all the variant records. Use the named scopes on [`ActiveStorage::Attachment`][] to avoid N+1s.

Reviewer: ```suggestion
If you're rendering lots of images on a page, the above example could result in N+1 queries loading all the variant records. To avoid these N+1 queries, use the named scopes on [`ActiveStorage::Attachment`][].
```

```ruby
user.avatars.with_all_variant_records.each do |file|

>   image_tag file.representation(resize_to_limit: [100, 100]).processed.url

Reviewer: This will process all N files on the first access right?

Author: Yes.

Reviewer: Should we maybe at a warning that it can make page slow?

Author: I guess maybe this isn't direct enough...

![image](https://user-images.githubusercontent.com/509837/122455829-71f47680-cf72-11eb-8dd1-03e637ce8ec4.png)

I could do followup.

Reviewer: This also bypasses HTTP caching, as it will create a new temporary url each time it's called.
So you need to either cache the url or users will download the file every time they access the page.
